hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/upgrade/BasicUpgradeFinalizer.java

```diff
@@ -436,4 +456,15 @@ private void logAndThrow(Exception e, String msg, ResultCodes resultCode)
   protected void updateLayoutVersionInDB(V vm, T comp) throws IOException {
     throw new UnsupportedOperationException();
   }
+
+  protected void postFinalizeUpgrade() throws IOException {
+  }
+
+  protected void finalizeVersionManager(Storage storageConfig)
+      throws UpgradeException {
+  }
+
+  protected boolean preFinalizeUpgrade() throws IOException {
```

Comment:
It seems like there is some overlap of logic between **preFinalizeUpgrade()** in various components, and org.apache.hadoop.ozone.upgrade.BasicUpgradeFinalizer#preFinalize. The preFinalize() step can be called either by the finalizer or the finalizationExecutor (based on whether we need that injection point), but in that implementation, we should be able to check the need to finalize as well as do pre-steps. If needed, we can split these 2 steps into needsFinalization() and preFinalize() steps, and call it from one place, the finalizationExecutor.

Reply:
I thought of moving it there and left it as it is to let the Specific DNFinalizer or SCMFinalizer decide as to what to do with pre/post/finalization functions.